src/MLIAP/mliap_descriptor_so3.cpp

```diff
@@ -0,0 +1,298 @@
+
+#define MAXLINE 1024
+#define MAXWORD 3
+
+
+MLIAPDescriptorSO3::MLIAPDescriptorSO3(LAMMPS *lmp,
+                                       char *paramfilename):
+  MLIAPDescriptor(lmp)
+{
+  nelements = 0;
+  elements = nullptr;
+  radelem = nullptr;
+  wjelem = nullptr;
+  so3ptr = nullptr;
+  read_paramfile(paramfilename);
+
+  so3ptr = new MLIAP_SO3(lmp,rcutfac,lmax,nmax,alpha);
+
+  ndescriptors = so3ptr->ncoeff;
+}
+
+MLIAPDescriptorSO3::~MLIAPDescriptorSO3()
+{
+  delete so3ptr;
+}
+
+void MLIAPDescriptorSO3::read_paramfile(char *paramfilename)
+{
+  int rcutfacflag = 0;
+  int nelementsflag = 0;
+  int elementsflag = 0;
+  int radelemflag = 0;
+  int wjelemflag = 0;
+  int nmaxflag=0;
+  int lmaxflag=0;
+  int alphaflag=0;
+
+  // set defaults for optional keywords
+
+  rfac0 = 0.99363;
+  rmin0 = 0.0;
+
+  // open SO3 parameter file on proc 0
+
+  FILE *fpparam;
+  if (comm->me == 0) {
+    fpparam = utils::open_potential(paramfilename,lmp,nullptr);
+    if (fpparam == nullptr)
+      error->one(FLERR,fmt::format
+        ("Cannot open SO3 parameter file {}: {}",
+          paramfilename, utils::getsyserror()));
+  }
+
+  char line[MAXLINE],*ptr;
+  int eof = 0;
+  int n,nwords;
+
+  while (1) {
+    if (comm->me == 0) {
+      ptr = fgets(line,MAXLINE,fpparam);
+      if (ptr == nullptr) {
+        eof = 1;
+        fclose(fpparam);
+      } else n = strlen(line) + 1;
+    }
+
+    MPI_Bcast(&eof,1,MPI_INT,0,world);
+    if (eof) break;
+    MPI_Bcast(&n,1,MPI_INT,0,world);
+    MPI_Bcast(line,n,MPI_CHAR,0,world);
+
+    // strip comment, skip line if blank
+
+    if ((ptr = strchr(line,'#'))) *ptr = '\0';
+    nwords = utils::count_words(line);
+    if (nwords == 0) continue;
+
+    // words = ptrs to all words in line
+    // strip single and double quotes from words
+
+    char* keywd = strtok(line,"' \t\n\r\f");
+    char* keyval = strtok(nullptr,"' \t\n\r\f");
+
+    if (comm->me == 0) {
+      utils::logmesg(lmp, fmt::format("SO3 keyword {} {} \n",
+        keywd, keyval));
+    }
+
+    // check for keywords with one value per element
+
+    if (strcmp(keywd,"elems") == 0 || strcmp(keywd,"radelems")
+      == 0 || strcmp(keywd,"welems") == 0) {
+
+      if (nelementsflag == 0 || nwords != nelements+1)
+        error->all(FLERR,"Incorrect SO3 parameter file");
+
+      if (strcmp(keywd,"elems") == 0) {
+        for (int ielem = 0; ielem < nelements; ielem++) {
+          elements[ielem] = utils::strdup(keyval);
+          keyval = strtok(nullptr,"' \t\n\r\f");
+        }
+
+        elementsflag = 1;
+      } else if (strcmp(keywd,"radelems") == 0) {
+        for (int ielem = 0; ielem < nelements; ielem++) {
+          radelem[ielem] = utils::numeric(FLERR,keyval,false,lmp);
+          keyval = strtok(nullptr,"' \t\n\r\f");
+        }
+        radelemflag = 1;
+      } else if (strcmp(keywd,"welems") == 0) {
+        for (int ielem = 0; ielem < nelements; ielem++) {
+          wjelem[ielem] = atof(keyval);
```

Comment:
Using `atof()` is dangerous since it does not check for a valid argument. At the very least this should use `utils::numeric()`, but even better would be to adapt the parser to use the aforementioned convenience classes, which do check for correctness.